File: Algorithms/104. Maximum Depth of Binary Tree/golang/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type NodeWDepth struct {
	node  *TreeNode
	depth int
}

type Stack []NodeWDepth

func (s Stack) Push(n NodeWDepth) Stack  { return append(s, n) }
func (s Stack) Pop() (Stack, NodeWDepth) { return s[:len(s)-1], s[len(s)-1] }
func (s Stack) NotEmpty() bool           { return len(s) > 0 }

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxDepth(root *TreeNode) int {
	s := make(Stack, 0)
	depth := 0
	var cur NodeWDepth
	s = s.Push(NodeWDepth{root, 0})

	for s.NotEmpty() {
		s, cur = s.Pop()

		if cur.node != nil {
			depth = max(depth, cur.depth+1)
			s = s.Push(NodeWDepth{cur.node.Right, cur.depth + 1})
			s = s.Push(NodeWDepth{cur.node.Left, cur.depth + 1})
		}
	}

	return depth
}

File: Algorithms/104. Maximum Depth of Binary Tree/javascript/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
    const stack = [];
    if (root) stack.push([1, root]);

    let depth = 0;
    while (stack.length) {
        let [ currentDepth, subRoot ] = stack.pop();
        if (subRoot) {
            depth = Math.max(depth, currentDepth);
            stack.push([currentDepth + 1, subRoot.left]);
            stack.push([currentDepth + 1, subRoot.right]);
        }
    }

    return depth;
};
